bitcoin/src/bip32.rs

```diff
@@ -256,108 +368,273 @@ impl serde::Serialize for ChildNumber {
+        /// Returns a [`DerivationPath`] from the [`NormalDerivationPath`].
+        pub fn to_derivation_path(&self) -> DerivationPath {
+            self.into_iter().map(|n| n.to_number()).collect()
+        }
 
-#[cfg(feature = "serde")]
-internals::serde_string_impl!(DerivationPath, "a BIP-32 derivation path");
+        /// Returns `true` if the [`NormalDerivationPath`] is empty.
+        pub fn is_empty(&self) -> bool { self.0.is_empty() }
 
-impl<I> Index<I> for DerivationPath
-where
-    Vec<ChildNumber>: Index<I>,
-{
-    type Output = <Vec<ChildNumber> as Index<I>>::Output;
+        /// Returns whether [`NormalDerivationPath`] represents master key (i.e. it's length
+        /// is empty). True for `m` path.
+        pub fn is_master(&self) -> bool { self.is_empty() }
 
-    #[inline]
-    fn index(&self, index: I) -> &Self::Output { &self.0[index] }
-}
+        /// Creates a new [`NormalDerivationPath`] that is a child of this one.
+        pub fn child(&self, cn: NormalChildNumber) -> Self { self.new_extended_from([cn]) }
 
-impl Default for DerivationPath {
-    fn default() -> DerivationPath { DerivationPath::master() }
-}
+        /// Converts into a [`NormalDerivationPath`] that is a child of this one.
+        pub fn into_with_child(self, cn: NormalChildNumber) -> Self {
+            let mut path = self;
+            path.extend_from([cn]);
+            path
+        }
+        /// Concatenates `path` to `self`.
+        ///
+        /// ```
+        /// use bitcoin::bip32::{NormalDerivationPath, NormalChildNumber};
+        /// use std::str::FromStr;
+        ///
+        /// let mut base = NormalDerivationPath::from_str("m/42").unwrap();
+        ///
+        /// base.extend_from(NormalDerivationPath::from_str("0/1").unwrap());
+        /// base.extend_from([
+        ///     NormalChildNumber::ZERO,
+        ///     NormalChildNumber::ONE
+        /// ]);
+        ///
+        /// assert_eq!(base, NormalDerivationPath::from_str("m/42/0/1/0/1").unwrap());
+        /// ```
+
+        pub fn extend_from<T: IntoIterator<Item = NormalChildNumber>>(&mut self, path: T) {
+            self.extend(path);
+        }
 
-impl<T> IntoDerivationPath for T
-where
-    T: Into<DerivationPath>,
-{
-    fn into_derivation_path(self) -> Result<DerivationPath, Error> { Ok(self.into()) }
-}
+        /// Concatenates `self` with `path` and return the resulting new path.
+        ///
+        /// ```
+        /// use bitcoin::bip32::{NormalDerivationPath, NormalChildNumber};
+        /// use std::str::FromStr;
+        ///
+        /// let base = NormalDerivationPath::from_str("m/42").unwrap();
+        ///
+        /// let deriv_1 = base.new_extended_from(NormalDerivationPath::from_str("0/1").unwrap());
+        /// let deriv_2 = base.new_extended_from([
+        ///     NormalChildNumber::ZERO,
+        ///     NormalChildNumber::ONE
+        /// ]);
+        ///
+        /// assert_eq!(deriv_1, deriv_2);
+        /// ```
+
```

Comment:
There is still an empty line here between the docstring and the function definition. Please fix this